modules/transactions/src/main/java/org/apache/ignite/internal/tx/impl/TxManagerImpl.java

```diff
@@ -253,12 +250,13 @@ public InternalTransaction begin(HybridTimestampTracker timestampTracker, boolea
 
     @Override
     public InternalTransaction begin(HybridTimestampTracker timestampTracker, boolean readOnly, TxPriority priority) {
-        HybridTimestamp beginTimestamp = clock.now();
-        UUID txId = transactionIdGenerator.transactionIdFor(beginTimestamp);
-        updateTxMeta(txId, old -> new TxStateMeta(PENDING, localNodeId, null, null, readOnly));
+        HybridTimestamp beginTimestamp = newBeginTimestamp(readOnly);
```

Comment:
It seems there is still a race. Let's imagine there are 2 threads: T1 starts a transaction, T2 checks whether there are not transactions started on versions strictly before 2.

1. Initially, all counters are 0
2. T1: Takes read lock, takes beginTs=now() [beginTs=10], releases the lock
3. T1: Resolves catalog version by beginTs, it's 1
4. T2: Takes write lock
5. T2: Takes now() (it's 12), resolves catalog version for 12, it's 2
6. T2: Sees that the version of interest (2) is already activated and all counters are 0, responds with TRUE
7. T1: Increments count for catalog version = 1 (which should never happen)

It seems that this can be fixed if the increment happens under read lock **together** with a check that the catalog version we just obtained is not less than some version for which we already responded 'yes' while handling a message.

Reply:
Fix it